File: cse_190_pa1/src/robot.cpp
#include "ros/ros.h"
#include "nav_msgs/Odometry.h"
#include "sensor_msgs/LaserScan.h"
#include <geometry_msgs/Twist.h>
#include "read_config.cpp"
#include <math.h>
#include <iostream>
#include <list>
#include "cse_190_assi_1/moveService.h"
#include "cse_190_assi_1/requestTexture.h"
#include "cse_190_assi_1/RobotProbabilities.h"
#include "cse_190_assi_1/temperatureMessage.h"
#include "std_msgs/Float32.h"
#include "std_msgs/Bool.h"
#include "std_msgs/String.h"


std::vector<std::vector<int> > motions;
std::vector<std::vector<float> > heat_map;
std::vector<std::vector<std::string> > pipe_map;
std::vector<std::vector<float> > probability_matrix;
Json::Value config; 
bool end_sim;
int motion_index;

ros::Subscriber tempSub;
ros::ServiceClient textClient;
ros::ServiceClient moveClient;
ros::Publisher temp_activator_pub;
ros::Publisher temp_res_pub;
ros::Publisher tex_res_pub;
ros::Publisher prob_res_pub;
ros::Publisher shutdown_pub;
//using namespace std;

void Print (const std::vector<std::vector<float>>& v){
  for (int i=0; i<v.size();i++){
  for (int j=0; j<v[0].size();j++){
    std::cout << v[i][j] << " ";
  }
    std::cout << std::endl;
  }
    std::cout << std::endl;
}

void Print (const std::vector<std::vector<int>>& v){
  for (int i=0; i<v.size();i++){
  for (int j=0; j<v[0].size();j++){
    std::cout << v[i][j] << " ";
  }
    std::cout << std::endl;
  }
    std::cout << std::endl;
}

void initialize_maps() {
        float temp_cold = 20.0;
        float temp_hot  = 40.0;
        float temp_warm = 25.0;

	pipe_map.resize(config["pipe_map"].size());
        for (int i=0; i<config["pipe_map"].size(); i++) {
	pipe_map[i].resize(config["pipe_map"][0].size());
            for (int j=0; j<config["pipe_map"][0].size(); j++) {
		pipe_map[i][j] = config["pipe_map"][i][j].asString();  
	    }
	}

	heat_map.resize(config["pipe_map"].size());
        for (int i=0; i<pipe_map.size(); i++) {
	heat_map[i].resize(config["pipe_map"][0].size());
            for (int j=0; j<pipe_map[0].size(); j++) {
                if (pipe_map[i][j] == "C")
                    heat_map[i][j] = temp_cold;
                else if (pipe_map[i][j] == "H")
                    heat_map[i][j] = temp_hot;
                else
                    heat_map[i][j] = temp_warm;
	    }
	}	

}

void initialize_beliefs() {
	float init_prob = 1.0/(float((config["texture_map"]).size())*float((config["texture_map"][0]).size()));
	probability_matrix.resize(config["pipe_map"].size());
        for (int i=0; i<pipe_map.size(); i++) {
	probability_matrix[i].resize(config["pipe_map"][0].size());
            for (int j=0; j<pipe_map[0].size(); j++) {
		probability_matrix[i][j] = init_prob;
	    }
	}
}

void update_prob_after_temp_measurement(float temp_reading) {
	std::vector<std::vector<float> > temp_probability_matrix(probability_matrix);
	float total_prob = 0.0;
	
        for (int i=0; i<probability_matrix.size(); i++) {
            for (int j=0; j<probability_matrix[0].size(); j++) {
		float prob_hit;
		if (config["temp_noise_std_dev"] == 0)
		   prob_hit = (heat_map[i][j] == temp_reading);
		else 
		   prob_hit = (1/(sqrt(2*M_PI)*config["temp_noise_std_dev"].asFloat()))*exp(-0.5*(pow(float(heat_map[i][j]-temp_reading),2.0)/pow(config["temp_noise_std_dev"].asFloat(),2.0))); 

		temp_probability_matrix[i][j] = prob_hit*probability_matrix[i][j];
		total_prob += temp_probability_matrix[i][j];
	    }
	}
  std::cout << "Update Temp Measurement " << motion_index << std::endl;
	Print(temp_probability_matrix);
	
        if (total_prob == 0)
            std::cout << "sum of probabilities is zero \n";
        else {
           for (int i=0; i<probability_matrix.size(); i++) {
               for (int j=0; j<probability_matrix[0].size(); j++) {
                  probability_matrix[i][j] = temp_probability_matrix[i][j]/total_prob;
	       }
	   }
	}

  std::cout << "Update Temp Measurement NORMALIZED" << motion_index << std::endl;
	Print(probability_matrix);
}

void update_prob_after_texture_measurement(std::string texture_reading) {
	std::vector<std::vector<float> > temp_probability_matrix(probability_matrix);
	float total_prob = 0.0;
	
        for (int i=0; i<probability_matrix.size(); i++) {
            for (int j=0; j<probability_matrix[0].size(); j++) {
		int hit;
		   hit = int(config["texture_map"][i][j] == texture_reading);
		   temp_probability_matrix[i][j] = (config["prob_tex_correct"].asFloat()*hit+(1-config["prob_tex_correct"].asFloat())*(1-hit))*probability_matrix[i][j];
		   total_prob += temp_probability_matrix[i][j];
	    }
	}
  std::cout << "Update Texture Measurement " << motion_index << std::endl;
	Print(temp_probability_matrix);
	
        if (total_prob == 0)
            std::cout << "sum of probabilities is zero \n";
        else {
           for (int i=0; i<probability_matrix.size(); i++) {
               for (int j=0; j<probability_matrix[0].size(); j++) {
                  probability_matrix[i][j] = temp_probability_matrix[i][j]/total_prob;
	       }
	   }
	}
  std::cout << "Update Texture Measurement NORMALIZED" << motion_index << std::endl;
	Print(probability_matrix);
}

void move_and_update(std::vector<int> curr_motion) {

	cse_190_assi_1::moveService srv;

	//srv.request.move = curr_motion;
	//std::copy(curr_motion.begin(), curr_motion.end(), srv.request.move);
	srv.request.move[0] = curr_motion[0];
	srv.request.move[1] = curr_motion[1];

	if (moveClient.call(srv)) {
  	     ROS_INFO("Succesfully called service moveService");
	} else {
  	     ROS_ERROR("Failed to call service moveService");
	}


	std::vector<std::vector<float> > temp_probability_matrix(probability_matrix);
	float total_prob = 0.0;
	int num_rows = probability_matrix.size();
	int num_cols = probability_matrix[0].size();

        for (int i=0; i<probability_matrix.size(); i++) {
            for (int j=0; j<probability_matrix[0].size(); j++) {
                   std::vector<std::vector<int>> possible_starts {{(i-1), j},{i, j},{(i+1), j},{i, (j-1)},{i, (j+1)}};

		   std::vector<int> correct_start = {(i-curr_motion[0]), (j-curr_motion[1])};
                   std::vector<std::vector<int> > incorrect_starts(possible_starts);
		   incorrect_starts.erase(std::remove(incorrect_starts.begin(), incorrect_starts.end(), correct_start),incorrect_starts.end());

  //std::cout << "i=" << i << " j=" << j << " Curr Motion: " << curr_motion[0] << " , " << curr_motion[1] << std::endl;
  //std::cout << "Correct Start: " << correct_start[0] << " , " << correct_start[1] << std::endl;
  //std::cout << "Incorrect Starts" << std::endl;
  //		   Print(incorrect_starts);
			   int correct_start_row = ((correct_start[0])%num_rows + num_rows)%num_rows;
			   int correct_start_col = ((correct_start[1])%num_cols + num_cols)%num_cols;
			
		   temp_probability_matrix[i][j] = (config["prob_move_correct"].asFloat())*probability_matrix[correct_start_row][correct_start_col];
		   for(int k=0; k<incorrect_starts.size(); k++) { 
			   int incorrect_start_row = ((incorrect_starts[k][0])%num_rows + num_rows)%num_rows;
			   int incorrect_start_col = ((incorrect_starts[k][1])%num_cols + num_cols)%num_cols;
			   //temp_probability_matrix[i][j] += ((1-(config["prob_move_correct"].asFloat()))/4)*temp_probability_matrix[incorrect_start_row][incorrect_start_col];
			   temp_probability_matrix[i][j] += ((1-(config["prob_move_correct"].asFloat()))/4)*probability_matrix[incorrect_start_row][incorrect_start_col];
		   }

		   total_prob += temp_probability_matrix[i][j];
	    }
	}
  std::cout << "Update Move Prob " << motion_index << std::endl;
	Print(temp_probability_matrix);	

        if (total_prob == 0)
            std::cout << "sum of probabilities is zero " << std::endl;
        else {
           for (int i=0; i<probability_matrix.size(); i++) {
               for (int j=0; j<probability_matrix[0].size(); j++) {
                  probability_matrix[i][j] = temp_probability_matrix[i][j]/total_prob;
	       }
	   }
	}
	
  std::cout << "Update Move Prob NORMALIZED" << motion_index << std::endl;
	Print(temp_probability_matrix);	
}

void on_temp_topic(const cse_190_assi_1::temperatureMessage::ConstPtr& msg) {
	
	// Read Temperature from topic 
	float temp_reading = msg->temperature;
  std::cout << "temp reading = " << temp_reading << std::endl;


	// Update prob after temperature measurement
	update_prob_after_temp_measurement(temp_reading);
	std_msgs::Float32 temp_res_msg;
	temp_res_msg.data = temp_reading; 
	temp_res_pub.publish(temp_res_msg);


	//Send service request to read texture
	std::string texture_reading;
	cse_190_assi_1::requestTexture srv;
	if (textClient.call(srv)) {
	     texture_reading = srv.response.data;
	} else {
  	     ROS_ERROR("Failed to call service requestTexture");
	}

  std::cout << "texture reading = " << texture_reading << std::endl;
	// Update prob after texture measurement
	update_prob_after_texture_measurement(texture_reading);
	std_msgs::String tex_res_msg;
	tex_res_msg.data = texture_reading; 
	tex_res_pub.publish(tex_res_msg);

        move_and_update(motions[motion_index++]);

	cse_190_assi_1::RobotProbabilities prob_msg;
	prob_msg.data.resize(probability_matrix.size()*probability_matrix[0].size());
	for(int i=0; i<probability_matrix.size(); i++) {
	   for(int j=0; j<probability_matrix[0].size(); j++) {
	   	   prob_msg.data[i*probability_matrix[0].size()+j] = probability_matrix[i][j];
	   }
	}
        //std::cout << "Final Probability Matrix \n" << probability_matrix << "\n";        
	prob_res_pub.publish(prob_msg);

        if (motion_index == motions.size()) {
        std::cout << "Motion complete\n";
  std::cout << "Motion index " << motion_index << std::endl;
	    end_sim = true; 
	}

}


int main(int argc, char **argv)
{
  ros::init(argc, argv, "robot");
  ros::NodeHandle n;
  config = read_config();
  std::cout << "Read Config Done" << std::endl;

  end_sim = false;
  motion_index = 0;
  //////////////////////////////////////////////
  // Define Publishers, Subscribers & Clients //
  //////////////////////////////////////////////

  // Subscribing for temperature topic
  tempSub = n.subscribe("/temp_sensor/data", 1, on_temp_topic);

 
  // Client for requestTexture service 
  textClient = n.serviceClient<cse_190_assi_1::requestTexture>("requestTexture");

  // Client for moveService 
  moveClient = n.serviceClient<cse_190_assi_1::moveService>("moveService");

  // Publisher for /temp_sensor/activation 
  temp_activator_pub = n.advertise<std_msgs::Bool>("/temp_sensor/activation", 10);

  // Publisher for  /results/temperature_data
  temp_res_pub = n.advertise<std_msgs::Float32>("/results/temperature_data", 10);

  // Publisher for  /results/texture_data
  tex_res_pub = n.advertise<std_msgs::String>("/results/texture_data", 10);

  // Publisher for  /results/probabilities
  prob_res_pub = n.advertise<cse_190_assi_1::RobotProbabilities>("/results/probabilities", 10);

  // Publisher for  /map_node/sim_complete
  shutdown_pub = n.advertise<std_msgs::Bool>("/map_node/sim_complete", 10);


  ////////////////////////////////////////////
  // Initialize heat maps & initial beliefs //
  ////////////////////////////////////////////

  initialize_maps();
  initialize_beliefs();

  std::cout << "Init maps & Beliefs Done" << std::endl;

  motions.resize(config["move_list"].size());
  //motions = config["move_list"];
  for (int i=0; i<config["move_list"].size(); i++) {
  	motions[i].resize(2);
  	motions[i][0] = config["move_list"][i][0].asInt();  
  	motions[i][1] = config["move_list"][i][1].asInt();  
  }

  std::cout << "Init Done" << std::endl;

  ros::Duration(2).sleep();

  ////////////////////// //////////////
  // Publish temp_activator_pub msg //
  ////////////////////////////////////

  std_msgs::Bool activation_msg;
  activation_msg.data = true;
  temp_activator_pub.publish(activation_msg);
  //ros::spin();
  ros::AsyncSpinner spinner(2);
  spinner.start();	

  while(end_sim == false) {
	// Run simulation
  ros::Duration(0.5).sleep();
  std::cout<<"Inside while testing End sim called" << std::endl;
  }

  std::cout<<"End sim called" << std::endl;
  
  activation_msg.data = false; 
  temp_activator_pub.publish(activation_msg);

  std_msgs::Bool shutdown_msg;
  shutdown_msg.data = true;
  shutdown_pub.publish(shutdown_msg);
  ros::Duration(1).sleep();
  ros::shutdown();

/*

  //ROS_INFO("WHEE");
  //ros::Duration(10).sleep();

  while (ros::ok())
  {
    geometry_msgs::Twist twist;   
    twist.linear.x = 10.0;
    twist.angular.z = 0.0;
    velocity_pub.publish(twist);

    ROS_INFO("Velocity: [%f %f]", twist.linear.x, twist.angular.z);

    ros::spinOnce();
  }
*/

  return 0;
}

File: cse_190_pa1/scripts/robot.py
#!/usr/bin/env python

import rospy
import random as r
import math as m
import numpy as np
from copy import deepcopy
from cse_190_assi_1.msg import temperatureMessage, RobotProbabilities
from cse_190_assi_1.srv import requestTexture, moveService
from std_msgs.msg import Bool, String, Float32
from collections import deque
from read_config import read_config


class RobotController():
    def __init__(self):
        self.config = read_config()
        rospy.init_node("robot_controller")
        self.temperature_subscriber = rospy.Subscriber(
                "/temp_sensor/data",
                temperatureMessage,
                self.handle_incoming_temperature_data
        )
        self.texture_requester = rospy.ServiceProxy(
                "requestTexture",
                requestTexture,
        )
        self.move_requester = rospy.ServiceProxy(
                "moveService",
                moveService,
        )
        self.temp_activator = rospy.Publisher(
                "/temp_sensor/activation",
                Bool,
                queue_size = 10
        )
        self.temp_res_pub = rospy.Publisher(
                "/results/temperature_data",
                Float32,
                queue_size = 10
        )
        self.tex_res_pub = rospy.Publisher(
                "/results/texture_data",
                String,
                queue_size = 10
        )
        self.prob_res_pub = rospy.Publisher(
                "/results/probabilities",
                RobotProbabilities,
                queue_size = 10
        )
        self.shutdown_pub = rospy.Publisher(
                "/map_node/sim_complete",
                Bool,
                queue_size = 10
        )
        self.initialize_maps()
        self.initialize_beliefs()
        self.motions = deque(self.config['move_list'])
        rospy.sleep(2)
        self.simulate()

    def initialize_maps(self):
        self.heat_map = self.generate_heatmap(self.config['pipe_map'])

    def initialize_beliefs(self):
        """
        set the initial probability matrix - everywhere is equally likely.
        """
        init_prob = 1.0 / float(len(self.config['texture_map'])) / float(len(self.config['texture_map'][0]))
        starting_beliefs = [[init_prob for row in range(len(self.config['texture_map'][0]))]
                            for col in range(len(self.config['texture_map']))]
        self.probability_matrix = starting_beliefs

    def generate_heatmap(self, pipe_map):
        """
        Generates a heat map based on whether there is a cold or hot pipe
        running under each tile. Tiles with no pipes underneath will have
        a "warm" temperature
        """
        temp_cold = 20.0
        temp_hot  = 40.0
        temp_warm = 25.0
        heat_map = [[ 0.0 for x in row] for row in pipe_map]
        for i in range(len(pipe_map)):
            for j in range(len(pipe_map[0])):
                if pipe_map[i][j] == 'C':
                    heat_map[i][j] = temp_cold

                elif pipe_map[i][j] == 'H':
                    heat_map[i][j] = temp_hot
                else:
                    heat_map[i][j] = temp_warm
        return heat_map

    def simulate(self):
        """
        Kicks the whole thing off. Calls rospy.spin() after activating
        the temperature sensor because from there on out everything happens
        on callback functions.
        """
        self.begin_simulation()
        rospy.spin()

    def begin_simulation(self):
        """
        Publishes to the topic that activates the temperature sensor.
        """
        activation_message = Bool()
        activation_message.data = True
        self.temp_activator.publish(activation_message)

    def end_simulation(self):
        """
        Publishes to the topic that stops the temperature sensor at the
        end of the simulation.
        """
        activation_message = Bool()
        activation_message.data = False
        self.temp_activator.publish(activation_message)
        self.shutdown_pub.publish(True)
        rospy.sleep(1)
        rospy.signal_shutdown("because I said so")

    def handle_incoming_temperature_data(self, message):
        """
        Callback function for the temp data. Initiates the timestep.
        """
        temp_reading = message.temperature
        self.timestep(temp_reading)

    def timestep(self, temp_reading):
        """
        Main sense->evalute->move function
        """
        self.update_from_temp(temp_reading)
        texture_reading = self.request_texture_reading()
        self.update_from_tex(texture_reading)
        self.publish_sensor_values(texture_reading, temp_reading)
        if len(self.motions) > 0:
            self.move_and_update(self.motions.popleft())
            self.publish_beliefs()
        else:
            print "Motion complete"
            self.show(self.probability_matrix)
            self.end_simulation()

    def publish_sensor_values(self, tex, temp):
        self.tex_res_pub.publish(tex)
        self.temp_res_pub.publish(temp)

    def publish_beliefs(self):
        prob_list = [p for row in self.probability_matrix for p in row]
        self.prob_res_pub.publish(prob_list)

    def show(self, p):
        """
        prints a probability matrix to the screen in a readable format
        """
        rows = ['[' + ','.join(map(lambda x: '{0:.5f}'.format(x),r)) + ']' for r in p]
        print '[' + ',\n '.join(rows) + ']'

    def update_from_temp(self, temp_reading):
        """
        Update the beliefs about the robot's position based on a temperature
        measurement.
        """
        temp_probs = deepcopy(self.probability_matrix)
        num_cols = len(temp_probs[0])
        num_rows = len(temp_probs)
        for i in range(num_rows):
            for j in range(num_cols):
                if self.config['temp_noise_std_dev'] == 0:
                    prob_hit = heat_map[i][j] == temp_reading
                else:
                    prob_hit = 1/(m.sqrt(2*m.pi)*self.config['temp_noise_std_dev'])*m.e**(-0.5*(float(self.heat_map[i][j] - temp_reading)/self.config['temp_noise_std_dev'])**2)
                temp_probs[i][j] = prob_hit * self.probability_matrix[i][j]
	print "Temperature Probabilities"
	self.show(temp_probs)

        total_prob = sum(sum(p) for p in temp_probs)
        if total_prob == 0:
            print "sum of probabilities is zero"
        else:
            temp_probs = [[prob/total_prob for prob in row] for row in temp_probs]
        self.probability_matrix = temp_probs
	print "Temperature Probabilities NORMALIZED"
	self.show(self.probability_matrix)
	#print self.probability_matrix

    def update_from_tex(self, texture_reading):
        """
        Update the beliefs about the robot's position based on a texture
        measurement.
        """
        temp_probs = deepcopy(self.probability_matrix)
        num_cols = len(temp_probs[0])
        num_rows = len(temp_probs)
        for i in range(num_rows):
            for j in range(num_cols):
                hit = int(self.config['texture_map'][i][j] == texture_reading)
                temp_probs[i][j] = (self.config['prob_tex_correct'] * hit + (1 - self.config['prob_tex_correct']) * (1 - hit)) * self.probability_matrix[i][j]
	print "Texture Probabilities"
	self.show(temp_probs)

        total_prob = sum(sum(p) for p in temp_probs)
        if total_prob == 0:
            print "sum of probabilities is zero"
        else:
            temp_probs = [[prob/total_prob for prob in row] for row in temp_probs]
        self.probability_matrix = temp_probs
	print "Texture Probabilities NORMALIZED"
	self.show(self.probability_matrix)

    def request_texture_reading(self):
        """
        Get a texture reading from the temperature sensor node via the
        texture request service.
        """
        response = self.texture_requester()
        texture_reading = response.data
        return texture_reading

    def move_and_update(self, move):
        self.send_move_command(move)
        self.update_beliefs_after_move(move)

    def send_move_command(self, move_command):
        self.move_requester(move_command)

    def update_beliefs_after_move(self, move_command):
        """
        Update position beliefs after attempting a move based on the known
        parameters of the motion model.
        """
        temp_probs = deepcopy(self.probability_matrix)

        num_cols = len(temp_probs[0])
        num_rows = len(temp_probs)
        for i in range(num_rows):
            for j in range(num_cols):
                possible_starts = [
                        [(i-1) % num_rows, j],
                        [i, j],
                        [(i+1) % num_rows, j],
                        [i, (j-1) % num_cols],
                        [i, (j+1) % num_cols]
                ]
                correct_start = [(i - move_command[0]) % num_rows, (j - move_command[1]) % num_cols]
                incorrect_starts = deepcopy(possible_starts)
                incorrect_starts.remove(correct_start)
                #print possible_starts, incorrect_starts, correct_start
                temp_probs[i][j] = self.config['prob_move_correct'] * self.probability_matrix[correct_start[0]][correct_start[1]]
                for start in incorrect_starts:
                    temp_probs[i][j] += ((1 - self.config['prob_move_correct'])/4) * temp_probs[start[0]][start[1]]
	print "Move Probabilities"
	self.show(temp_probs)

        total_prob = sum(sum(p) for p in temp_probs)
        if total_prob == 0:
            print "sum of probabilities is zero"
        else:
            temp_probs = [[prob/total_prob for prob in row] for row in temp_probs]
        self.probability_matrix = temp_probs
	print "Move Probabilities NORMALIZED"
	self.show(self.probability_matrix)


if __name__ == '__main__':
    rc = RobotController()
